pulumi/src/backupCron.ts

```diff
@@ -0,0 +1,110 @@
+import { ServiceConfig, StatefulService } from ".";
+import * as k8s from '@pulumi/kubernetes'
+
+export const deployStsBackupCron = (asset: string, sts: StatefulService, namespace: string, provider: k8s.Provider) => {    
```

Comment:
breaking things out into separate functions has it's uses, but depending on the types accepted by those functions, you can lose useful type narrowing along the way. I would personally just have a single func (assuming we can get serviceAccountName) as follows:

```ts
export const deployStsBackupCron = (asset: string, sts: StatefulService, namespace: string, provider: k8s.Provider) => {    
  if (!sts.backup) return

  const pvcs = sts.services.map((svc, i) => `data-${svc.name}-${asset}-sts-${i}`).join(',')

  const backupContainer = {
    name: `${asset}-backup-runner`,
    image: 'lukmyslinski/backuprunner:0.24',
    args: ['-n', namespace, '-s', `${asset}-sts`, '-p', pvcs, '-r', `${sts.replicas}`, "-c", `${sts.backup.backupCount}`],
  }

  new k8s.batch.v1.CronJob(`${asset}-backup-job`, {
    metadata: {
      name: `${asset}-backup-job`,
      namespace: namespace,
      annotations: { 'pulumi.com/skipAwait': 'true' },
    },
    spec: {
      successfulJobsHistoryLimit: 1,
      failedJobsHistoryLimit: 1,
      concurrencyPolicy: "Forbid",
      schedule: sts.backup.schedule,
      jobTemplate: {
        spec: {
          template: {
            spec: {
              serviceAccountName: 'passThroughArg',
              containers: [backupContainer],
              restartPolicy: "Never"
            },
          },
        },
      },
    },
  }, { provider })
}
```

Reply:
no need for `!!` or `?` as we have already narrowed the type and retain that scope.

we will need to build out `sts.backup` a bit more. we can either require `backupCount` and `schedule` in which we will want to validate in `pulumi/config.ts`, or we can change those to optional types and provide defaults internally:

```ts
  if (!sts.backup) return

  const { backupCount = 1, schedule = '0 4 * * 0' } = sts.backup
  
  OR

  const backupCount = sts.backup.backupCount ?? 1
  const schedule = sts.backup.schedule ?? '0 4 * * 0' 
```

Reply:
I see, definitely makes sense to do so. I'm more accustomed to stricter SRP coming mostly from the JVM land but having fewer types to verify is definitely a huge gain here

Reply:
And just to clarify, does the config validation make sense regarding the backup config? Let me know if not and I can walk through a bit more, might be easier to do a quick screen share than try to explain via text.

Reply:
You mean the scoped type narrowing via `if` guards? All clear

Reply:
It does enforce a certain code structure but it's liveable. Not as good as i.e. a monadic approach but hey, it's simple enough

Reply:
more from the pulumi config side. see: https://github.com/shapeshift/unchained/blob/d13399f8a1d3f782f891bbde423cd5a3410c3d88/pulumi/src/config.ts

`new Pulumi.Config` does not enforce types unfortunately, so we have extra validation in there to basically say "if this optional object is defined, ensure all required fields actually exists" so we don't end up with an `undefined` on something we think should be required based on the type def